File: client/src/services/donation.ts
// @ts-nocheck
import { StellarService } from './stellar';
import { CharityService } from './charity';

export interface DonationDetails {
  amount: string;
  charityId: string;
  donorId: string;
  message?: string;
}

export class DonationService {
  private static instance: DonationService;
  private stellarService: StellarService;
  private charityService: CharityService;

  private constructor() {
    this.stellarService = StellarService.getInstance();
    this.charityService = CharityService.getInstance();
  }

  static getInstance(): DonationService {
    if (!DonationService.instance) {
      DonationService.instance = new DonationService();
    }
    return DonationService.instance;
  }

  async makeDonation(
    fromSecretKey: string,
    details: DonationDetails
  ): Promise<string> {
    try {
      const charity = await this.charityService.getCharityById(details.charityId);
      if (!charity) {
        throw new Error('Charity not found');
      }

      // Make the donation using Stellar
      const txHash = await this.stellarService.makeDonation(
        fromSecretKey,
        charity.walletAddress,
        details.amount,
        details.message || 'Thank you for your donation!'
      );

      // TODO: Store donation record in your backend
      // TODO: Update charity.totalDonations

      return txHash;
    } catch (error) {
      console.error('Error processing donation:', error);
      throw new Error('Failed to process donation');
    }
  }

  async getDonationsByCharity(charityId: string): Promise<any[]> {
    // TODO: Implement fetching donations for a charity from your backend
    return [];
  }

  async getDonationsByUser(userId: string): Promise<any[]> {
    // TODO: Implement fetching donations made by a user from your backend
    return [];
  }
} 

File: server/models/donation.go
package models

import (
	"gorm.io/gorm"
)

type Donation struct {
	gorm.Model
	Amount    string  `json:"amount"`
	CharityID uint    `json:"charityId"`
	DonorID   string  `json:"donorId"`
	Message   string  `json:"message"`
	TxHash    string  `json:"txHash" gorm:"unique"`
	Status    string  `json:"status"`
	Category  string  `json:"category"`
	Charity   Charity `json:"charity" gorm:"foreignKey:CharityID"`
	Donor     User    `json:"donor" gorm:"foreignKey:FirebaseID"`
}
